### ISSUE TITLE
Weaknesses in the `Metadata` Identification Method

### ISSUE THREAD

**Issue by Hellobloc** [blockscout/blockscout-rs#416]
### Background
As mentioned in the documentation with blockscout, your approach to distinguish `Main Part` and `Metadata Part` is to validate the metadata part by compiling the `contract` and `compiler_modified` separately to look for different content. 

This is clever because it allows contracts containing multiple metadata to be validated, and it could in theory even support validation of the following contracts.
```solidity 
pragma solidity ^0.8.0;
contract Deploy2 {
    address public immutable a;

    bytes32 public ptr;
    constructor(address d) public {
        a = address(d);//
    }
}
contract Deploy3 {

    bytes32 public ptr;//

}
contract A {
    bytes public constant code = type(Deploy2).creationCode;
    bytes public constant code2 = type(Deploy3).creationCode;
    bytes1 public immutable flag;
    bytes1 public immutable flag2;
    constructor() public {
        flag = code[code.length-20]&0xaa;
        flag2 = code2[code2.length-20]&0xaa;
    }
}
```

But only if `compiler_modified` does not trigger changes to other non-metadata parts.
Your design for `compiler_modified` is as follows, i.e., it causes changes to the metadata by adding a `Library` that does not seem to be used.
https://github.com/blockscout/blockscout-rs/blob/2cdb3c2c8e532c3d8ac3718cf01a3462d36636c1/smart-contract-verifier/smart-contract-verifier/src/verifier/contract_verifier.rs#L100-L121
### Risk Description
However, this seems to cause the `Library` address to be marked as `metadata` in some extreme cases, which may pose some risks, such as the `library` address and other adjacent bytecodes provided in the source code not corresponding to the `real bytecode`.

## Exploit
The user can make the `library` address also at risk of being marked as `metadata` by adding a contract with `SOME_TEXT_USED_AS_FILE_NAME` as the file name in the contract source code and defining a `library` named `L_` in that file.
The details are as follows：
```javascript
var solc = require('solc');

var input = {
          language: 'Solidity',
          settings: {
            libraries: {
              'SOME_TEXT_USED_AS_FILE_NAME': {
                L_: '0x4200000000000000000000000000000000000001'
              }
            },
            outputSelection: {
              '*': {
                '*': ['evm.bytecode']
              }
            }
          },
          sources: {
            'SOME_TEXT_USED_AS_FILE_NAME': {
              content: 'library L_ { function f() public returns (uint) { return 7; } }'
            },
            'a.sol': {
              content: 'import "SOME_TEXT_USED_AS_FILE_NAME"; contract A { function g() public { L_.f(); } }'
            }
          }
        };

var output = JSON.parse(solc.compile(JSON.stringify(input)));
for (var contractName in output.contracts['a.sol']) {
  console.log(
    contractName +
      ': ' +
      output.contracts['a.sol'][contractName].evm.bytecode.object
  );
}
//A: 60806040...4200000000000000000000000000000000000001...0033

```
The following code demonstrates that when such a contract occurs, your code will likely cause the library address to change, which may cause the library to be marked as metadata as well.
We hope the following simple code example can help us communicate this problem.
```rust
#![allow(unused)]
fn main() {
use std::collections::BTreeMap;
use std::ops::Add;
use std::ops::Bound::Included;
use std::path::{Path, PathBuf};



let mut libraries = BTreeMap::new();
let mut lib1 = BTreeMap::new();
let mut lib2 = BTreeMap::new();
lib1.insert("Library1", "0x000000000");
lib1.insert("Library2", "0x000000000");
lib1.insert("Library3", "0x000000000");
lib1.insert("Library4", "0x000000000");


lib2.insert("L_", "0x000000000");


libraries.insert(PathBuf::from("1.sol"),lib1.clone());
libraries.insert(PathBuf::from("SOME_TEXT_USED_AS_FILE_NAME"),lib2.clone());




//It's the same as blockscout
let lib_entry = libraries.entry(PathBuf::from("SOME_TEXT_USED_AS_FILE_NAME")).or_default();
let non_used_contract_name = lib_entry
                .keys()
                .map(|key| key.chars().next().unwrap_or_default().to_string())
                .collect::<Vec<_>>()
                .join("_")
                .add("_");
println!("Unused Contract name : {non_used_contract_name}");
lib_entry.insert(&non_used_contract_name,"0xcafecafecafecafecafecafecafecafecafecafe",);
println!("Final Result:");
for (library_name, address) in lib_entry {
    println!("{library_name} = {address}");
}
}
//Unused Contract name : L_
//Final Result:
//L_ = 0xcafecafecafecafecafecafecafecafecafecafe
```

## Impact
This problem may cause **the library addresses** and **other adjacent bytecodes that make up the cbor code** in the malicious source code to differ from the real code. Causes of fraud
https://github.com/blockscout/blockscout-rs/blob/2865ae1cbbc7a6f18bf8e2b237fdd656535f3239/smart-contract-verifier/smart-contract-verifier/src/verifier/bytecode.rs#L247-L321
## Recommendation
We recommend using another way to cause the `metadata` to be different, rather than using a `library` that may cause `bytecode` inconsistencies.

Finally, the metadata tagging method you designed for the contract source system is truly ingenious. And thank you very much for the open source of your project to further my understanding of this specific process.

**Comment by Hellobloc** [comment]
@sevenzing

**Comment by Hellobloc** [comment]
Because you have some cbor parsing of the metadata, we may also need to ensure here that Library can form a cbor encoding format alone or with other adjacent bytecodes.

**Comment by Hellobloc** [comment]
@rimrakhimov

**Comment by rimrakhimov** [comment]
@Hellobloc First of all, thank you for raising this issue. You've done an excellent job researching how our verification algorithm works and delving deep into the details! We're curious if this research is for a specific project or just personal interest? Regardless, great work, and thank you very much! :+1:

Regarding the issue you've described, I see what you are talking about. We use only the first letters of possible libraries located in SOME_TEXT_USED_AS_FILE_NAME, likely to ensure that the final library name doesn't become too long, and that causes some problems. At the same time, the libraries metadata field was chosen for modifications because it's a parameter common to all contracts being compiled.

A potential workaround here might be to modify the `sources` input field to include an additional suffix in the file names. The modified input for the example you provided would look like:
```
{
    "language": "Solidity",
    "settings": {
      "libraries": {
        "SOME_TEXT_USED_AS_FILE_NAME": {
          "L_": "0x4200000000000000000000000000000000000001"
        }
      },
      "outputSelection": {
        "*": {
          "*": ["evm.bytecode.object"]
        }
      }
    },
    "sources": {
      "SOME_TEXT_USED_AS_FILE_NAME.mod": {
        "content": "library L_ { function f() public returns (uint) { return 7; } }"
      },
      "a.sol.mod": {
        "content": "import \"SOME_TEXT_USED_AS_FILE_NAME\"; contract A { function g() public { L_.f(); } }"
      }
    }
  }
```
Please let us know if you see any issues with this approach.

Additionally, just interested, have you managed to exploit the described weakness to verify a contract deployed on any of our Blockscout chains, or if it remains a theoretical possibility for now?

**Comment by Hellobloc** [comment]
I'm sorry that I haven't taken the time to reproduce the problem because the work related to this issue has kept me busy, 

I just pulled out some of your code snippets to do a simple verification of this attack, but I think the likelihood of this attack happening is high because I don't currently see any code content that would prevent said attack.

The only thing to note is that the library will probably need to form a cbor code with the previous bytecode, which shouldn't be too much of a problem considering that we can use assignment constants to make arbitrary content appear in the bytecode. 
For example, the following
```solidity
bytes32 x = 0xa118ccc25901df;
B.F();
```

**Comment by Hellobloc** [comment]
And do you mind if we add a telegram for further communication, because my current research needs to be communicated with an expert like you who has developed a very good source code verification tool！

**Comment by Hellobloc** [comment]
This is my Tg : lucas_block

**Comment by rimrakhimov** [comment]
@Hellobloc I'll write you a little later

**Comment by Hellobloc** [comment]
> @Hellobloc I'll write you a little later. But in general I prefer discussions inside the Github ecosystem, as they are more transparent

Thank you very much, I also like the Github discussion of these, but because there are some issues on our side is not convenient to discuss publicly, look forward to communicate with you on TG.

**Comment by Hellobloc** [comment]
https://eth-goerli.blockscout.com/address/0x2606695009cC0f639Aa1D34462Fe29CE5B91a4c6?tab=contract

**Comment by Hellobloc** [comment]
I have constructed a case of source code inconsistency on my side, which you can see below, but please be sure to keep this address information, as we are working on it and such pages are helpful.
I think for this issue fix we need to discuss it and look forward to your tg message.

**Comment by Hellobloc** [comment]
I recently constructed an attack case, and from what I've seen so far his impact is still quite large and can lead to differences in bytecode as well as library addresses.

You can see the variables a,b in the source code of the following contract and their values.
![image](https://user-images.githubusercontent.com/113746997/231546454-57174fe6-85cb-44b9-b1fc-6bafd3e3652a.png)
![image](https://user-images.githubusercontent.com/113746997/231546496-73be93e2-5ba9-4353-abb2-2cf93ce99b9b.png)

https://eth-goerli.blockscout.com/address/0xB3085B77E24798d79D2fAF8495B20d570753d86F?tab=contact_code

**Comment by Hellobloc** [comment]
I think you might consider giving me a public acknowledgement and a financial reward or something, haha. Just kidding! @rimrakhimov

**Comment by rimrakhimov** [comment]
Closed in #420
